src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,123 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_LENGTH = 10;
+    private T[] data;
+    private int size;
+
+    public ArrayList() {
+        data = (T[]) new Object[DEFAULT_LENGTH];
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        arrayIncrease(size + 1);
+        data[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        arrayIncrease(size + 1);
+        if (checkIndex(index) || index == size) {
+            System.arraycopy(data, index, data, index + 1, size - index);
+            data[index] = value;
+            size++;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index is not exists");
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int temp = size;
+        for (int i = 0; i < list.size(); i++) {
+            arrayIncrease(size + 1);
+            data[i + temp] = list.get(i);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (checkIndex(index)) {
+            return data[index];
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index is not exists");
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (checkIndex(index)) {
+            data[index] = value;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index is not exists");
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (checkIndex(index)) {
+            T value = data[index];
+
+            if (index == data.length - 1) {
+                data[index] = null;
+            } else {
+                System.arraycopy(data, index + 1, data, index, size - index);
+            }
+            size--;
+
+            return value;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index is not exists");
+        }
     }
 
     @Override
-    public T remove(T element) {
-        return null;
+    public T remove(T element) throws NoSuchElementException {
+        int index = findIndex(element);
+        if (index != -1) {
+            System.arraycopy(data, index + 1, data, index, size - index);
+            size--;
+        } else {
+            throw new NoSuchElementException("Index is not exists");
```

Comment:
Error message here is a bit misleading, because we are trying to remove element, but no an index.